cuda/trsv_kernels.cuh

```diff
@@ -0,0 +1,1096 @@
+               global_col < mtx.length(1))
+                ? mtx(global_row, global_col)
+                : ar_type{0};
+    }
+    group.sync();
+    // Invert lower triangular matrix
+
+    // First: make diagonals 1
+    if (dmtx == dmtx_t::non_unit) {
+        if (threadIdx.y == 0) {
+            for (int row = 0; row < swarp_size; ++row) {
+                const int col = threadIdx.x;
+                const auto triang_idx = col * triang_stride + row;
+                const auto local_val = triang[triang_idx];
+                const auto diag_el = ar_type{1} / swarp.shfl(local_val, row);
+                if (col <= row) {
+                    triang[triang_idx] =
+                        (row == col) ? diag_el : local_val * diag_el;
+                }
+            }
+        }
+    }
+    // No need to sync since the same warp will continue to compute
+
+    if (threadIdx.y == 0) {
+        // Compute remaining inverse in-place (with Gauss-Jordan elimination)
+        for (int diag_id = 0; diag_id < swarp_size; ++diag_id) {
+            const int col = threadIdx.x;
+            const auto diag_el = triang[diag_id * triang_stride + diag_id];
+            for (int row = diag_id + 1; row < swarp_size; ++row) {
+                auto factor =
+                    -triang[diag_id * triang_stride + row];  // broadcast
+                const int triang_idx = col * triang_stride + row;
+                const auto triang_val = triang[triang_idx];
+                if (col < row) {
+                    triang[triang_idx] =
+                        (col == diag_id)
+                            ? factor * diag_el
+                            : triang_val +
+                                  factor *
+                                      triang[col * triang_stride + diag_id];
+                }
+                swarp.sync();
+            }
+        }
+    }
+    group.sync();
+
+    constexpr index_type num_local_rows = swarp_size / swarps_per_block;
+    ar_type local_row_result[num_local_rows] = {};
+    for (index_type col_block = 0; col_block < row_block_idx; ++col_block) {
+        const index_type global_col = col_block * swarp_size + threadIdx.x;
+
+        // Wait until result is known for current column block
+        if (threadIdx.x == 0 && threadIdx.y == 0) {
+            // Note: this needs to be signed since the initial value is
+            //       ~0 (all ones)
+            volatile auto *last_col = reinterpret_cast<std::make_signed_t<
+                std::remove_pointer_t<decltype(col_row_global_helper)>> *>(
+                col_row_global_helper);
+            while (*last_col < col_block) {
+            }
+        }
+        group.sync();
+        // Make sure that data is only read after synchronization
+        __threadfence();
+        const auto x_cached = x(global_col, 0);
+        const index_type end_local_row_idx =
+            (mtx.length(0) - 1 - row_block_idx * swarp_size - threadIdx.y) /
+            swarps_per_block;
```

Comment:
I wonder whether it will leave some rows in the end not processed.
For example, mtx.lenght(0) = 14, row_block_idx = 0, swarp_size = 32, swarps_per_block = 4
process 3, 3, 2, 2 is not enough

Reply:
I know this code part is counter-intuitive, but I needed to write it this way so the loop is properly unrolled. If I did the check with `global_row` (which I originally did) the performance is a lot worse.

The idea is that consecutive warps in a thread block should read consecutive rows. Here, each warp in a thread block (indexed with `threadIdx.y`) reads `start + threadIdx.y + N * swarps_per_block` (with `N < 32/swarps_per_block`).

In your example, `end_local_row_idx == (13 - threadIdx.y) / 4`
| threadIdx.y | end_local_row_idx | last accepted global_row |
| ------------ | -------------------- | --------------------------- |
| 0 | 3 | 12 |
| 1 | 3 | 13 |
| 2 | 2 | 10 |
| 3 | 2 | 11 |

So it should work fine for your example (I also ran this example in DEBUG mode to see if I get out of bound or a mismatch with cuSparse, but it was fine).